File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        most = 0
        chars = Counter()

        left = right = 0

        while right < len(s):
            r = s[right]
            chars[r] += 1
            while chars[r] > 1:
                chars[s[left]] -= 1
                left += 1
            most = max(most, right - left + 1)
            right += 1
        return most

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> counter;
        int left = 0;
        int right = 0;
        int longest = 0;
        for (int right = 0; right < s.size(); ++right) {
            ++counter[s[right]];
            if (counter[s[right]] > 1) {
                while (counter[s[right]] > 1) {
                    --counter[s[left++]];
                }
            }

            longest = std::max(longest, right - left + 1);
        }

        return longest;
    }
};